Volumetric locking solid

sveerara opened the pull request:
- Edited Stress Divergence kernels for B-bar correction.
- Edited linear small strain formulations (Linear.C, PlaneStrain.C and Axisymmetric.C) to include b-bar correction. 
- Updated tests

Closes  #7260

sveerara committed:
New code changes with correct jacobian and volumetric locking as an option #7260

sveerara committed:
Correction to tests for volumetric locking change #7260

dschwen commented:
Bison fails with this. Do you have a patch ready?

dschwen commented on modules/solid_mechanics/src/kernels/StressDivergenceRZ.C:
```diff
@@ -41,6 +43,21 @@ StressDivergenceRZ::StressDivergenceRZ(const InputParameters & parameters)
 {
+  // calculate volume averaged value of shape function derivative
+  Real Avg_derivative_0 = 0.0; // Average shape function derivatives in direction 0
+  Real Avg_derivative_1 = 0.0; // Average shape function derivatives in direction 1
```
This should be formulated using a RealVector value. Then you do not need to do component wise math below.

dschwen commented on modules/solid_mechanics/src/kernels/StressDivergenceRZ.C:
```diff
@@ -41,6 +43,21 @@ StressDivergenceRZ::StressDivergenceRZ(const InputParameters & parameters)
+  {
+    Avg_derivative_0 += _grad_test[_i][qp_i](0) * _JxW[qp_i] * _coord[qp_i];
+    Avg_derivative_1 += _grad_test[_i][qp_i](1) * _JxW[qp_i] * _coord[qp_i];
+    element_vol += _JxW[qp_i] * _coord[qp_i];
```
You do not calculate this yourself there is `_current_elem->volume()`

dschwen commented on modules/solid_mechanics/src/materials/AxisymmetricRZ.C:
```diff
@@ -51,6 +52,36 @@ AxisymmetricRZ::computeStrain( const unsigned qp,
+  Real trace = strain_increment.trace();
+  strain_increment.xx() += volumetric_strain - trace / dim;
+  strain_increment.yy() += volumetric_strain - trace / dim;
+  strain_increment.zz() += volumetric_strain - trace /dim;
```
spacing

dschwen commented on modules/solid_mechanics/src/materials/Linear.C:
```diff
@@ -62,8 +63,39 @@ Linear::computeStrain( const unsigned qp,
+  {
+    volumetric_strain += (_grad_disp_x[qp_loop](0) + _grad_disp_y[qp_loop](1) + _grad_disp_z[qp_loop](2)) / 3.0 * _solid_model.JxW(qp_loop);
+
+    volume += _solid_model.JxW(qp_loop);
```
you may want to cache the result of `_solid_model.JxW(qp_loop)`

dschwen commented:
Ok, I _really_ would like to see a somewhat more verbose description of of code change that makes it necessary to re-gold **over 260 tests**!

sveerara commented:
In the new commits, I have edited the Jacobian to be consistent with the volumetric locking change. I have also added a flag "volumertic_locking_correction" using which volumetric locking can be turned in the strain calculation and stress divergence calculation. 

I have edited the NonlinearRZ.C and Axisymmetric.C files to use the correct volume (consistent with coord = RZ) for averaging the volumetric strain.

sveerara commented:
Bison tests are failing. I have a patch ready for this.

dschwen commented:
Could you squash this into one code commit and one test commit, please.

sveerara commented:
I have squashed the commits into one code commit and one test commit. Thanks!

dschwen commented:
Looks like you picked up one of Cody's commits.

sveerara commented:
Thanks! Now, only my commits are in the pull request.

dschwen commented:
In a separate commit could you add the single element Jacobian tests with locking correction? You mentioned that you already have those. I think it'd be good to have them in the test suite.

sveerara committed:
Single element tests to verify jacobian with volumetric locking correction #7260

sveerara commented:
I have added the single element tests for small strain linear, RZ and planestrain formulations. 

For the finite strain versions, the jacobian will not match exactly. I have added the expected error in jacobian when the corresponding finite strain versions are used.

bwspenc commented:
I'm working with Swetha to run the BISON assessment cases and resolve a few issues that this causes. We should have this sorted out in the next day or so.  I think this is about ready to go, but hold off on merging this until I let you know that we have resolved those issues.

bwspenc commented:
We have the BISON regression and assessment case fixes ready to go, so this is ready to be merged.

dschwen commented on modules/solid_mechanics/src/kernels/StressDivergenceRZ.C:
```diff
@@ -67,6 +123,75 @@ StressDivergenceRZ::computeQpResidual()
+
+        vol += _JxW[_qp] * _coord[_qp];
+      }
+      _avg_grad_test[_i][_component] /= vol;
```
use current_elem->volume() everywhere

sveerara replied:
I have used _current_elem_volume instead of _current_elem->volume(). 
_current_elem->volume() does not contain the coordinate system correction.

dschwen commented on modules/solid_mechanics/src/kernels/StressDivergenceRZ.C:
```diff
@@ -101,11 +227,103 @@ StressDivergenceRZ::calculateJacobian( unsigned int ivar, unsigned int jvar )
+    else
+      val += (tmp.xx() + tmp.yy() + tmp.zz()) * (_avg_grad_test[_i][1] - new_test(1)) / 3.0;
+  }
   return val;
```
remove all `/ 3.0` above and add a single `/ 3.0` on the return line

sveerara replied:
There are four terms that are contributing to val. Only two of them have / 3.0. So I cannot make this change on the return line.

dschwen replied:
Huh? What am I missing. I'm seeing a factor 1/3 in all terms :-/

sveerara replied:
The term val is calculated starting from line 225. The term on line 225 does not have a factor of 1/3. The term on line 241 has a factor of 1/9. The other two terms are calculated in the lines 245-260 and only these two terms have a factor of 1/3.

dschwen replied:
A factor of 1/9 is 1/3 * 1/3, so that factor is in there. The
initialization could be made to zero with the addition of the single term
happening in the return statement. It is no big deal. The change would
remove one division operation and remove a little bit of clutter. If you
prefer to leave it like it is that's fine, too.

On Wed, Sep 7, 2016 at 8:02 PM sveerara <notifications@github.com> wrote:

> In modules/solid_mechanics/src/kernels/StressDivergenceRZ.C
> <https://github.com/idaholab/moose/pull/7531#discussion_r77935924>:
>
> > +    if ((ivar == 0) && (jvar == 0))
> > +      val += (sum_3x1(0) * test.xx() +  sum_3x1(2) * test.zz()) * (_avg_grad_phi[_j][0] - new_phi(0)) / 3.0;
> > +    else if ((ivar == 0) && (jvar == 1))
> > +      val += (sum_3x1(0) * test.xx() +  sum_3x1(2) * test.zz()) * (_avg_grad_phi[_j][1] - new_phi(1)) / 3.0;
> > +    else if ((ivar == 1) && (jvar == 0))
> > +      val += sum_3x1(1) * test.yy() * (_avg_grad_phi[_j][0] - new_phi(0)) / 3.0;
> > +    else
> > +      val += sum_3x1(1) * test.yy() * (_avg_grad_phi[_j][1] - new_phi(1)) / 3.0;
> > +
> > +    // Bvol^T_i * C * B_j
> > +    // tmp = C * B_j from above
> > +    if (ivar == 0)
> > +      val += (tmp.xx() + tmp.yy() + tmp.zz()) * (_avg_grad_test[_i][0] - new_test(0)) / 3.0;
> > +    else
> > +      val += (tmp.xx() + tmp.yy() + tmp.zz()) * (_avg_grad_test[_i][1] - new_test(1)) / 3.0;
> > +  }
> >    return val;
>
> The term val is calculated starting from line 225. The term on line 225
> does not have a factor of 1/3. The term on line 241 has a factor of 1/9.
> The other two terms are calculated in the lines 245-260 and only these two
> terms have a factor of 1/3.
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/7531/files/2461c4a5d32f7ff6d53d6a3619b3f5f971929d67#r77935924>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMWPv7AonAaZ93Vctr_Yce9iOLBhL0Vks5qn2yggaJpZM4Jinz_>
> .
>

sveerara replied:
I have made this change. Thanks!

dschwen commented on modules/solid_mechanics/src/kernels/StressDivergenceRZ.C:
```diff
@@ -101,11 +227,103 @@ StressDivergenceRZ::calculateJacobian( unsigned int ivar, unsigned int jvar )
+      Real vol = 0.0;
+      // calculate volume averaged value of shape function derivative
+      _avg_grad_test.resize(_test.size());
+      for (_i = 0; _i < _test.size(); _i++)
```
would you mind switching to pre-increment in for loops everywhere? (`; ++_i)`). It's never slower and sometimes faster (plus it is semantically what you _really_ mean)

friedmud replied:
Unless it has "been decided" and is codified in the style guidelines I don't like this advice.  We have tons of post-increment for counters (it's how I increment simple counters for instance).  No reason to change this unless we're going to do it everywhere and enforce it.

dschwen replied:
It was a loose suggestion because it is the _de facto_ standard in tensor mechanics and phase field with >95% of all for loops being incremented this way. I'd like to avoid introducing code that shifts away from this.  Of course this code only goes into solid mechanics, so this doesn't matter much. Just a heads up for when this is implemented in TM.

sveerara replied:
I will make this change in the tensor mechanics version.

dschwen commented on modules/solid_mechanics/src/kernels/StressDivergenceRZ.C:
```diff
@@ -101,11 +227,103 @@ StressDivergenceRZ::calculateJacobian( unsigned int ivar, unsigned int jvar )
+          else
+            _avg_grad_test[_i][_component] += _grad_test[_i][_qp](_component) * _JxW[_qp] * _coord[_qp];
+
+          vol += _JxW[_qp] * _coord[_qp];
```
see above (use _current-elem->volume())

dschwen commented on modules/solid_mechanics/src/materials/Nonlinear3D.C:
```diff
@@ -73,26 +74,30 @@ Nonlinear3D::computeIncrementalDeformationGradient( std::vector<ColumnMajorMatri
+    for ( unsigned qp=0; qp < _solid_model.qrule()->n_points(); ++qp )
+    {
+      const Real det_Fhat( detMatrix( Fhat[qp] ) );
+      const Real factor( std::pow( det_Fhat_average/det_Fhat, third ) );
```
replace with `std::cbrt`, it is three times faster!

dschwen commented on modules/solid_mechanics/src/materials/NonlinearPlaneStrain.C:
```diff
@@ -94,25 +95,30 @@ NonlinearPlaneStrain::computeIncrementalDeformationGradient( std::vector<ColumnM
+    for ( unsigned qp=0; qp < _solid_model.qrule()->n_points(); ++qp )
+    {
+      const Real det_Fhat( detMatrix( Fhat[qp] ) );
+      const Real factor( std::pow( det_Fhat_average/det_Fhat, third ) );
```
same here

dschwen commented on modules/solid_mechanics/src/materials/NonlinearRZ.C:
```diff
@@ -72,25 +73,30 @@ NonlinearRZ::computeIncrementalDeformationGradient( std::vector<ColumnMajorMatri
+    for ( unsigned qp=0; qp < _solid_model.qrule()->n_points(); ++qp )
+    {
+      const Real det_Fhat( detMatrix( Fhat[qp] ) );
+      const Real factor( std::pow( det_Fhat_average/det_Fhat, third ) );
```
and here

dschwen commented:
Good point about the `elem->volume()` I did not realize it does not contain the coordinate system correction! I'll have to go over the places where I used it and make sure it does not break on non-cartesian systems.

permcody commented:
If anyone has any ideas for dealing with this volume issue, I'd like to hear them. I suppose the problem is that libMesh isn't aware of our coordinate system and this is a pure libMesh function. Do we need to get heavy handed with this in some way to prevent user error?

dschwen commented:
Is a Grizzly fix ready, too?

sveerara commented:
I dint know that Grizzly is failing too. @bwspenc can you please check what is failing in Grizzly? I do not have access to Grizzly or to the App tests results. Thanks!

bwspenc commented:
I didn't realize it broke that test.  I'll get a fix ready ASAP.

bwspenc commented:
I just submitted a PR to grizzly with a fix for that test.

dschwen commented:
I'll go ahead and merge this now. Please update the App PRs and merge them soon.
